### PR TITLE
Improve Kafka non happy flow(s).

### PR DESCRIPTION
**Pull request by PremyslCerny**
I let ChatGPT to analyze the Kafka code.

### CODE REVIEW

File: bspump/kafka/source.py [hunk 1]
```diff
@@ -151,17 +155,15 @@ async def main(self):
 						self.LastRefreshTopicsTime = current_time
```
Comment (mejroslav) [inline bspump/kafka/source.py:155]: What is the reason? ChatGPT?😀
This has to be strongly tested before it is merged…

Comment (PremyslCerny) [reply bspump/kafka/source.py:155]: Which line you are commenting? :D

Comment (PremyslCerny) [reply bspump/kafka/source.py:155]: Only the green or red ones are changed.

File: bspump/kafka/source.py [hunk 2]
```diff
@@ -151,17 +155,15 @@ async def main(self):
...
 						break
 
-					m = c.poll(0.2)
+					m = await self.ProactorService.execute(c.poll, 0.5)  # Increased poll interval
```
Comment (mejroslav) [inline bspump/kafka/source.py:158]: @PremyslCerny Sorry for the previous comment - this is what I am talking about.
Why did you increased the poll interval?
I suggest to at least make it a global variable and add a comment about why this value is used.
Or make it configurable so we can easily test different values and find the optimal parameterization.

File: bspump/kafka/source.py [hunk 3]
```diff
@@ -151,17 +155,15 @@ async def main(self):
...
 
 					if m is None:
 						await asyncio.sleep(self.Sleep)
 						continue
 
 					if m.error():
-						L.error("The following error occured while polling for messages: '{}'.".format(m.error()))
+						L.error("The following error occurred while polling for messages: '{}'.".format(m.error()))
```
Comment (mejroslav) [inline bspump/kafka/source.py:165]: Can you include some struct data like a topic name or something?

**Review by mejroslav — CHANGES_REQUESTED** [review]

File: bspump/kafka/source.py (hunk 2 above)
Comment (mejroslav) [reply bspump/kafka/source.py:180]: @PremyslCerny 

```python
m = await self.ProactorService.execute(c.poll, self.PollInterval)
```

Comment (PremyslCerny) [reply bspump/kafka/source.py]: I added comments

**Review by mejroslav — APPROVED** [review]
One nitpick, otherwise approved. The default value of poll interval should be further optimized after testing.

**Comment by PremyslCerny** [comment]
I implemented the buffering of consumed messages, as discussed. With the help of ChatGPT.

**Review by ateska — COMMENTED** [review]
Ok, please test a throughput - see the other Kafka MR, I left the same request there.

**Comment by PremyslCerny** [comment]
This is not behaving very well when multiple topics are being read from. We need to work more explicitly with commits / offset storage.

File: bspump/kafka/source.py [hunk 4]
```diff
@@ -123,49 +158,78 @@ def __init__(self, app, pipeline, connection, id=None, config=None):
...
+			if m is None:
+				continue
+
+			if m.error():
+				L.error("The following error occurred while polling for messages: '{}'.".format(m.error()))
+				consumer.unsubscribe()
 				return
 
-			c.subscribe(self.Subscribe)
+			self.Buffer.append(m)
```
Comment (ateska) [inline bspump/kafka/source.py:198]: Buffer has to be locked here.

File: bspump/kafka/source.py [hunk 5]
```diff
@@ -123,49 +158,78 @@ def __init__(self, app, pipeline, connection, id=None, config=None):
...
-			try:
+			if len(self.Buffer) >= self.BufferSize or (current_time - self.LastFlushTime) > self.BufferTimeout:
 
-				while 1:
-					await self.Pipeline.ready()
-					current_time = self.App.time()
+				try:
+					future = asyncio.run_coroutine_threadsafe(self.flush_buffer(consumer), self.Loop)
+					# Wait for the result:
+					future.result()
```
Comment (ateska) [inline bspump/kafka/source.py:205]: This looks strange, honestly.

**Review by ateska — CHANGES_REQUESTED** [review]

File: bspump/kafka/source.py (hunk 4 above)
Comment (PremyslCerny) [reply bspump/kafka/source.py:198]: There is only one thread accessing the buffer at a time.

File: bspump/kafka/source.py (hunk 5 above)
Comment (PremyslCerny) [reply bspump/kafka/source.py:205]: I know, but this is how it works and they mention it in their official docs: https://docs.python.org/3/library/asyncio-dev.html

File: bspump/kafka/source.py [hunk 6]
```diff
@@ -123,49 +158,78 @@ def __init__(self, app, pipeline, connection, id=None, config=None):
...
+		This method runs on a thread and fills the buffer with the polled messages from the Kafka consumer.
+
+		Running Kafka's poll method on a separate thread in an asynchronous application enhances responsiveness
+		by freeing the main thread for other tasks. It ensures non-blocking processing, improving throughput and
+		scalability, as message fetching and processing are decoupled.
+		This approach also optimizes resource utilization by leveraging multiple CPU cores and enhances error
+		handling, making the application more resilient and maintainable.
+		"""
 		while self.Running:
+			current_time = self.App.time()
```
Comment (ateska) [inline bspump/kafka/source.py:180]: Don't use App.time() in threads, this is an event loop bound method.

File: bspump/kafka/source.py [hunk 7]
```diff
@@ -174,9 +238,76 @@ async def main(self):
...
+			except BaseException as e:
+				L.exception("Error when connecting to Kafka")
+				self.Pipeline.set_error(None, None, e)
+				return
+
+			c.subscribe(self.Subscribe)
+
+			try:
+				# Run the poll loop on a separate thread
+				await self.ProactorService.execute(self.poll_kafka, c)
```
Comment (ateska) [inline bspump/kafka/source.py:295]: Don't use ProactorService for this, launch your own thread.

Also consider (for performance reasons) launching more than one consumer thread - in that case, `self.Buffer` concept will have to be extended.

File: bspump/kafka/source.py [hunk 8]
```diff
@@ -123,49 +158,78 @@ def __init__(self, app, pipeline, connection, id=None, config=None):
...
+			if m.error():
+				L.error("The following error occurred while polling for messages: '{}'.".format(m.error()))
+				consumer.unsubscribe()
 				return
 
-			c.subscribe(self.Subscribe)
+			self.Buffer.append(m)
 
-			try:
+			if len(self.Buffer) >= self.BufferSize or (current_time - self.LastFlushTime) > self.BufferTimeout:
```
Comment (ateska) [inline bspump/kafka/source.py:200]: You have to this check also when `m is None` b/c of the timeout ...

**Review by ateska — CHANGES_REQUESTED** [review]

**Comment by PremyslCerny** [comment]
You definitely not making it easy for me :D ;)

**Comment by PremyslCerny** [comment]
@ateska Are there other remarks? I would like to test only when we agree on the design.

**Review by ateska — CHANGES_REQUESTED** [review]
What is the max. throughput of this?

Design seems to be ok - I'm a bit worried about impact of these locks and passing buffers along.
It has to be tested since it can have an impact on the design.

File: bspump/kafka/source.py [hunk 9]
```diff
@@ -121,50 +161,93 @@ def __init__(self, app, pipeline, connection, id=None, config=None):
...
+		by freeing the main thread for other tasks. It ensures non-blocking processing, improving throughput and
+		scalability, as message fetching and processing are decoupled.
+		This approach also optimizes resource utilization by leveraging multiple CPU cores and enhances error
+		handling, making the application more resilient and maintainable.
+		"""
+		while self.Running:
+			current_time = time.time()
 
-	async def main(self):
+			if self.RefreshTopics > 0 and current_time > self.LastRefreshTopicsTime + self.RefreshTopics:
```
[the remark this answers is not in the document]
Comment (PremyslCerny) [reply bspump/kafka/source.py:193]: Yes

**Comment by PremyslCerny** [comment]
We can discuss it tomorrow, the maximum throughput is exactly the same as with the previous version aroun 100k on my machine, with parsec it is also the same about 4k (but this is maybe due to the infrastructure limits).

**Comment by ateska** [comment]
> We can discuss it tomorrow, the maximum throughput is exactly the same as with the previous version aroun 100k on my machine, with parsec it is also the same about 4k (but this is maybe due to the infrastructure limits).

100k of what? :-)

**Pull request merged**